### PR TITLE
dotCMS/core#26899 [UI] When image loading place holder loose focus, the placeholder stays in the editor. 

### LINKED ISSUE #26899: [UI] When image loading place holder loose focus, the placeholder stays in the editor.
### Parent Issue
#26505

### User Story
When the image placeholder while waiting for the AI endpoint lose focus, the placeholder stays in the editor. 
 
image

https://github.com/dotCMS/core/assets/31667212/fdf6b485-d25c-4bb6-b432-0f7509e48f05

### Acceptance Criteria

- The placeholder should be replaced with the actual image form AI. 
- We can't trust in the current selection from tiptap to perfom operations in a node. 

### dotCMS Version
master

### Proposed Objective
Quality Assurance

### Proposed Priority
Priority 2 - Important

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.
_to_define_

### Assumptions & Initiation Needs
_to_define_

### Quality Assurance Notes & Workarounds
_to_define_

### Sub-Tasks & Estimates
_to_define_

### PR DESCRIPTION
**Pull request by hmoreras**
### Proposed Changes
* Fix the duplication of image / loader, when the user move the focus of the editor. 
* Don't trust on editor `selection` to manipulated the image generated by AI. 
* Inlcude a property in the data `AI_IMAGE_PLACEHOLDER_PROPERTY ` to identify when an image is a placeholder, while the user confirm is want the image generated. 
* Add the option of insert the `LoaderNode ` in a specific position. 

#### Before 

https://github.com/dotCMS/core/assets/31667212/fdf6b485-d25c-4bb6-b432-0f7509e48f05

#### After 

https://github.com/dotCMS/core/assets/31667212/bd48996c-dd6a-4d5a-844a-acd6cec340e4

### CODE REVIEW

**Review by oidacra — APPROVED** [review]

File: core-web/libs/block-editor/src/lib/extensions/ai-content-prompt/plugins/ai-content-prompt.plugin.ts [hunk 1]
```diff
@@ -148,15 +148,15 @@ export class AIContentPromptView {
                 filter((deleteContent) => deleteContent)
             )
             .subscribe(() => {
-                const nodeInformation: DotTiptapNodeInformation = findNodeByType(
+                const nodeInformation: DotTiptapNodeInformation[] = findNodeByType(
```
Comment (zJaaal) [inline core-web/libs/block-editor/src/lib/extensions/ai-content-prompt/plugins/ai-content-prompt.plugin.ts:151]: do you need the array? I see that you are always accessing the first index, should we destructure the array here or access the first index here right away?

Comment (hmoreras) [reply core-web/libs/block-editor/src/lib/extensions/ai-content-prompt/plugins/ai-content-prompt.plugin.ts:151]: I changed from single to multiple, just to make the function more generic. and eventually re use it for the image-placeholder search.

Comment (hmoreras) [reply core-web/libs/block-editor/src/lib/extensions/ai-content-prompt/plugins/ai-content-prompt.plugin.ts]: done

File: core-web/libs/block-editor/src/lib/extensions/ai-content-prompt/plugins/ai-content-prompt.plugin.ts [hunk 2]
```diff
@@ -115,10 +115,9 @@ export class AIContentPromptView {
                 filter((state) => state.acceptContent)
             )
             .subscribe((state) => {
-                const nodeInformation: DotTiptapNodeInformation = findNodeByType(
-                    this.editor,
-                    NodeTypes.AI_CONTENT
-                );
+                const [nodeInformation] = findNodeByType(this.editor, NodeTypes.AI_CONTENT) || [
+                    null
+                ];
```
Comment (rjvelazco) [inline core-web/libs/block-editor/src/lib/extensions/ai-content-prompt/plugins/ai-content-prompt.plugin.ts:120]:
```suggestion
                const nodeInformation = findNodeByType(this.editor, NodeTypes.AI_CONTENT)?.[0];
```

File: core-web/libs/block-editor/src/lib/extensions/ai-content-prompt/plugins/ai-content-prompt.plugin.ts [hunk 3]
```diff
@@ -148,10 +147,9 @@ export class AIContentPromptView {
                 filter((deleteContent) => deleteContent)
             )
             .subscribe(() => {
-                const nodeInformation: DotTiptapNodeInformation = findNodeByType(
-                    this.editor,
-                    NodeTypes.AI_CONTENT
-                );
+                const [nodeInformation] = findNodeByType(this.editor, NodeTypes.AI_CONTENT) || [
+                    null
+                ];
```
Comment (rjvelazco) [inline core-web/libs/block-editor/src/lib/extensions/ai-content-prompt/plugins/ai-content-prompt.plugin.ts:152]:
```suggestion
                const nodeInformation = findNodeByType(this.editor, NodeTypes.AI_CONTENT)?.[0];
```

File: core-web/libs/block-editor/src/lib/extensions/ai-image-prompt/plugins/ai-image-prompt.plugin.ts [hunk 4]
```diff
@@ -104,21 +123,28 @@ export class AIImagePromptView {
...
-                    .run();
+                const loaderNodes = findNodeByType(this.editor, NodeTypes.LOADER);
+
+                //Trust in this property to identify the image as a placeholder, until the user accept the content.
+                data[AI_IMAGE_PLACEHOLDER_PROPERTY] = true;
+
+                if (loaderNodes) {
+                    this.editor
+                        .chain()
+                        .deleteRange({ from: loaderNodes[0].from, to: loaderNodes[0].to })
```
Comment (rjvelazco) [inline core-web/libs/block-editor/src/lib/extensions/ai-image-prompt/plugins/ai-image-prompt.plugin.ts:134]: We can get the first one from above

File: core-web/libs/block-editor/src/lib/shared/utils/prosemirror.utils.ts [hunk 5]
```diff
@@ -244,33 +244,49 @@ export const replaceNodeWithContent = (
...
+ * @param {Editor} editor - The TipTap editor instance.
+ * @returns {DotTiptapNodeInformation | null}
+ */
+export const getAIPlaceholderImage = (editor: Editor): DotTiptapNodeInformation => {
+    const nodes = findNodeByType(editor, NodeTypes.DOT_IMAGE);
+    const aIPlaceholderImages = nodes
+        ? nodes.filter((nodeInfo) => nodeInfo.node.attrs.data[AI_IMAGE_PLACEHOLDER_PROPERTY])
+        : null;
+
+    return aIPlaceholderImages?.length ? aIPlaceholderImages[0] : null;
```
Comment (rjvelazco) [inline core-web/libs/block-editor/src/lib/shared/utils/prosemirror.utils.ts:289]:
```suggestion
    return aIPlaceholderImages?.[0]
```

File: core-web/libs/block-editor/src/lib/extensions/ai-image-prompt/plugins/ai-image-prompt.plugin.ts [hunk 6]
```diff
@@ -104,21 +123,28 @@ export class AIImagePromptView {
...
+                        .insertImage(data, loaderNodes[0].from)
+                        .openAIContentActions(DOT_AI_IMAGE_CONTENT_KEY)
```
Comment (rjvelazco) [inline core-web/libs/block-editor/src/lib/extensions/ai-image-prompt/plugins/ai-image-prompt.plugin.ts:136]: We can centralize all Extension Keys in a Map. What are your thoughts, @hmoreras and @oidacra?

**P.S:** Not for this PR though.

File: core-web/libs/block-editor/src/lib/nodes/ai-content/ai-content.node.ts [hunk 7]
```diff
@@ -70,7 +70,9 @@ export const AIContentNode = Node.create({
                 (content?: string) =>
                 ({ commands, editor, tr }) => {
                     // Check if the AI_CONTENT node exists in the document.
-                    const nodeInformation = findNodeByType(editor, NodeTypes.AI_CONTENT);
+                    const [nodeInformation] = findNodeByType(editor, NodeTypes.AI_CONTENT) || [
+                        null
+                    ];
```
Comment (rjvelazco) [inline core-web/libs/block-editor/src/lib/nodes/ai-content/ai-content.node.ts:75]:
```suggestion
                    const nodeInformation = findNodeByType(editor, NodeTypes.AI_CONTENT)?.[0]
```

File: core-web/libs/block-editor/src/lib/nodes/image-node/image.node.ts [hunk 8]
```diff
@@ -99,7 +99,8 @@ export const ImageNode = Image.extend({
                     const { head } = selection;
                     const node = {
                         attrs: getImageAttr(attrs),
-                        type: ImageNode.name
+                        type: ImageNode.name,
+                        storage: { aiStatus: 'pending' }
```
Comment (rjvelazco) [inline core-web/libs/block-editor/src/lib/nodes/image-node/image.node.ts:103]: Is there an `Enum` for this?

Comment (hmoreras) [reply core-web/libs/block-editor/src/lib/nodes/image-node/image.node.ts:103]: is not, and actually that need to be removed.

**Review by zJaaal — APPROVED** [review]

**Review by rjvelazco — APPROVED** [review]

**Pull request merged**